File: matlab/dynamics.m
function dxdt = dynamics(x, u, mr, mw, l, Iw)
    g = 9.81;
    theta = x(1);
    dtheta = x(2);
    dphi = x(4);

    dxdt = zeros(4, 1);
    dxdt(1) = dtheta;
    dxdt(2) = (((1/2)*mr*g*l + mw*g*l)*sin(theta) - u)/((1/3)*mr*l^2 + mw*l^2);
    dxdt(3) = dphi;
    dxdt(4) = u/Iw - dxdt(2);
end

File: components/logic/dynamics/dynamics.c
#include "dynamics.h"
#include <math.h>

void nonlinear_continuous(const state_t * x, const float u, const dynamics_params_t * params, state_t * dxdt)
{
  float theta, dtheta, dphi, mr, mw, l, Iw, g;
  mr = params->mr;
  mw = params->mw;
  l = params->l;
  Iw = params->Iw;
  g = params->g;
  theta = x->theta;
  dtheta = x->dtheta;
  dphi = x->dphi;

  dxdt->theta = dtheta;
  dxdt->dtheta = ((0.5f * mr * g * l + mw * g * l) * sinf(theta) - u) / ((1.0f / 3.0f) * mr * l * l + mw * l * l);
  dxdt->phi = dphi;
  dxdt->dphi = u / Iw - dxdt->dtheta;
  return;
}